### ISSUE TITLE
Issue of scan self aborting aborting(?) and stop() or abort() not putting RE in stopped state.

### ISSUE THREAD

**Issue by ambarb** [NSLS-II/Bug-Reports#129]
Exiting blue sky was how to resolve.

User in the field encountered this problem.

```
In [356]: fccd_set(1,7200)
Current fCCD config: 7200.0 images @ 1.00 Hz = 7200.000 s

In [357]: RE(ct())
Transient Scan ID: 66794
Persistent Unique Scan ID: '4890899b-c4f3-4124-8751-64ecdfc642cc'
logit Measurement #66794 @ 705.75 eV, 1.00 Hz, at 105.0 K for 7200 images over 7200 s                  
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|   seq_num |       time | fccd_stats1_total | fccd_stats2_total | fccd_stats3_total | fccd_stats4_total | fccd_stats5_total |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|         1 | 13:23:47.6 |            469832 |           3754983 |          14925068 |           2038328 |       49119544171 |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
generator ct ['489089'] (scan num: 66794)
Out[357]: ['4890899b-c4f3-4124-8751-64ecdfc642cc']

In [358]: logit Measurement #66794 @ 705.75 eV, 1.00 Hz, at 105.0 K for 7200 images over 7200 s

In [359]: fccd_set(20,7200)
Current fCCD config: 144000.0 images @ 20.00 Hz = 7200.000 s

In [360]: fccd_set(1,7200)
Current fCCD config: 7200.0 images @ 1.00 Hz = 7200.000 s

In [361]: fccd_set(1,60)
Current fCCD config: 60.0 images @ 1.00 Hz = 60.000 s

In [362]: ct_all_dark
Out[362]: <function __main__.ct_all_dark>

In [363]: RE(ct_all_dark())

Starting procedure to acquire camera gains at 1.0Hz

Camera stopped..

Gain bit set to 2

Camera restarted..
Transient Scan ID: 66795
Persistent Unique Scan ID: 'af476d14-bfc3-42a0-95ac-a3436055c294'
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|   seq_num |       time | fccd_stats1_total | fccd_stats2_total | fccd_stats3_total | fccd_stats4_total | fccd_stats5_total |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|         1 | 13:29:47.0 |             65685 |            580283 |            397931 |            220201 |       49091154518 |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
generator ct ['af476d'] (scan num: 66795)

Camera stopped..

Gain bit set to 1

Camera restarted..
Transient Scan ID: 66796
Persistent Unique Scan ID: 'e0351a1b-4df6-4846-b1bd-3fc87b8bbc05'
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|   seq_num |       time | fccd_stats1_total | fccd_stats2_total | fccd_stats3_total | fccd_stats4_total | fccd_stats5_total |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|         1 | 13:31:19.1 |        -574166388 |       -4921269824 |       -4919591161 |       -2296951233 |       33975201017 |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
generator ct ['e0351a'] (scan num: 66796)

Camera stopped..

Gain bit set to 0

Camera restarted..
Transient Scan ID: 66797
Persistent Unique Scan ID: '41ec3f8b-a635-4515-a345-96913091d0d0'
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|   seq_num |       time | fccd_stats1_total | fccd_stats2_total | fccd_stats3_total | fccd_stats4_total | fccd_stats5_total |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
|         1 | 13:32:50.9 |       -1724645803 |      -14782016165 |      -14771036324 |       -6901125377 |        3718692607 |
+-----------+------------+-------------------+-------------------+-------------------+-------------------+-------------------+
generator ct ['41ec3f'] (scan num: 66797)

Returned to initial gain setting and restarting acquisition
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/home/xf23id1/Users/2016_07_Roopali/20160715_overnighttoFridayam.py in <module>()
----> 1 RE(ct_all_dark())

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, raise_if_interrupted, **metadata_kw)
    560         self._metadata_per_call.update(metadata_kw)
    561 
--> 562         gen = ensure_generator(plan)
    563 
    564         self._plan_stack.append(gen)

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/plan_tools.py in ensure_generator(plan)
     28     gen : coroutine
     29     """
---> 30     gen = iter(plan)  # no-op on generators; needed for classes
     31     if not isinstance(gen, PLAN_TYPES):
     32         # If plan does not support .send, we must wrap it in a generator.

TypeError: 'NoneType' object is not iterable

In [364]: fccd_set(20,7200)
Current fCCD config: 144000.0 images @ 20.00 Hz = 7200.000 s

In [365]: fccd_set(20,3600)
Current fCCD config: 72000.0 images @ 20.00 Hz = 3600.000 s

In [366]: fccd_set(1,7200)
Current fCCD config: 7200.0 images @ 1.00 Hz = 7200.000 s

In [367]: fccd_set(20,3600)
Current fCCD config: 72000.0 images @ 20.00 Hz = 3600.000 s

In [368]: fccd_set(10,3600)
Current fCCD config: 36000.0 images @ 10.00 Hz = 3600.000 s

In [369]: fccd_set(20,3600)
Current fCCD config: 72000.0 images @ 20.00 Hz = 3600.000 s

In [370]: fccd_set(10,3600)
Current fCCD config: 36000.0 images @ 10.00 Hz = 3600.000 s

In [371]: fccd_set(10,7200)
Current fCCD config: 72000.0 images @ 10.00 Hz = 7200.000 s

In [372]: fccd_set(10,7200)
Current fCCD config: 72000.0 images @ 10.00 Hz = 7200.000 s

In [373]: fccd_set(1,7200)
Current fCCD config: 7200.0 images @ 1.00 Hz = 7200.000 s

In [374]: fccd_set(10, 360)
Current fCCD config: 3600.0 images @ 10.00 Hz = 360.000 s

In [375]: RE
RE

In [375]: RE(ct())
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/home/xf23id1/Users/2016_07_Roopali/20160715_overnighttoFridayam.py in <module>()
----> 1 RE(ct())

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, raise_if_interrupted, **metadata_kw)
    537         # First thing's first: if we are in the wrong state, raise.
    538         if not self.state.is_idle:
--> 539             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    540 
    541         futs = []

RuntimeError: The RunEngine is in a running state

In [376]: 
KeyboardInterrupt

In [376]: 
KeyboardInterrupt

In [376]: RE(ct())
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/home/xf23id1/Users/2016_07_Roopali/20160715_overnighttoFridayam.py in <module>()
----> 1 RE(ct())

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, raise_if_interrupted, **metadata_kw)
    537         # First thing's first: if we are in the wrong state, raise.
    538         if not self.state.is_idle:
--> 539             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    540 
    541         futs = []

RuntimeError: The RunEngine is in a running state

In [377]: RE(ct())
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/home/xf23id1/Users/2016_07_Roopali/20160715_overnighttoFridayam.py in <module>()
----> 1 RE(ct())

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, raise_if_interrupted, **metadata_kw)
    537         # First thing's first: if we are in the wrong state, raise.
    538         if not self.state.is_idle:
--> 539             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    540 
    541         futs = []

RuntimeError: The RunEngine is in a running state

In [378]: stop()
Stopping...

In [379]: fccd_set(10, 360)
Current fCCD config: 3600.0 images @ 10.00 Hz = 360.000 s

In [380]: RE(ct())
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/home/xf23id1/Users/2016_07_Roopali/20160715_overnighttoFridayam.py in <module>()
----> 1 RE(ct())

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, raise_if_interrupted, **metadata_kw)
    537         # First thing's first: if we are in the wrong state, raise.
    538         if not self.state.is_idle:
--> 539             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    540 
    541         futs = []

RuntimeError: The RunEngine is in a running state

In [381]: 

In [381]: 

In [381]: abort()
Aborting....
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
/home/xf23id1/Users/2016_07_Roopali/20160715_overnighttoFridayam.py in <module>()
----> 1 abort()

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/global_state.py in abort()
    155     This function closes over the global RunEngine instance, gs.RE.
    156     """
--> 157     return gs.RE.abort()
    158 
    159 

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in abort(self, reason)
    745         self._reason = reason
    746         self._exception = RequestAbort()
--> 747         self._task.cancel()
    748         for task in self._failed_status_tasks:
    749             task.cancel()

AttributeError: 'NoneType' object has no attribute 'cancel'

In [382]: RE(ct())
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
/home/xf23id1/Users/2016_07_Roopali/20160715_overnighttoFridayam.py in <module>()
----> 1 RE(ct())

/home/xf23id1/conda_envs/collection/lib/python3.4/site-packages/bluesky/run_engine.py in __call__(self, plan, subs, raise_if_interrupted, **metadata_kw)
    537         # First thing's first: if we are in the wrong state, raise.
    538         if not self.state.is_idle:
--> 539             raise RuntimeError("The RunEngine is in a %s state" % self.state)
    540 
    541         futs = []

RuntimeError: The RunEngine is in a running state

In [383]: 
```

**Comment by ambarb** [comment]
`RE(ct_all_dark())` should be just `ct_all_dark()`.  I am not sure how you guys want to handle this - i.e. don't do this or if there is something you can do to fix things.  However, data might be compromised.  We had to retake images.

**Comment by danielballan** [comment]
I gather that `ct_all_dark()` is a function that calls `RE` and returns `None`. So, if everything is working as it should, a mistake like `RE(ct_all_dark())` should:
1. Successfully run `ct_all_dark()` with data saved.
2. Attempt to run `RE(None)` (because `ct_all_dark()` returns `None`).
3. Raise an exception, because `None` is not a valid plan.
4. Leave `RE` in an idle state, ready to try again.

It looks like everything except (4) worked properly. From what I can see here, I do not think your data was compromised. A fairly recent fix to bluesky should have addressed this. I am not sure whether we deployed that fix to CSX yet. Can you check `bluesky.__version__`?

**Comment by ambarb** [comment]
conda list:
bluesky                   0.5.3                    py34_0 

Looking are their data, they can access it.  There was a problem with it not related to the way they were scanning.

Here is the definition of `ct_all_dark()`

```
def ct_all_dark ():

    print('\nStarting procedure to acquire camera gains at {}Hz'.format(1/fccd.acquire_time.get()))
    caput('XF:23ID1-OP{FBck}Sts:FB-Sel',0) #maintain finepitch 
    sleep(2)
    # Save current settings
    shutter_open = shchk()
    initial_gain = caget('XF:23ID1-ES{FCCD}cam1:FCRICGain')

    # Close shutter
    shclose()

    for gain_bit in [2, 1, 0]: # ALL GAINS corresponds to camera gain of [1, 2, 8] respectively
    #for gain_bit in [0]: # ONLY AUTO GAIN 
        #caput('XF:23ID1-ES{FCCD}cam1:CameraPwr',0)
        time.sleep(5)
        caput('XF:23ID1-ES{FCCD}cam1:Acquire', 0)
        print('\nCamera stopped..')
        time.sleep(5)
        caput('XF:23ID1-ES{FCCD}cam1:FCRICGain', gain_bit)#, wait=False)
        print('\nGain bit set to {}'.format(gain_bit))
        time.sleep(5)
        caput('XF:23ID1-ES{FCCD}cam1:Acquire', 1)
        time.sleep(5)
        #caput('XF:23ID1-ES{FCCD}cam1:CameraPwr',1)
        print('\nCamera restarted..')
        time.sleep(5)
        _ct_dark()

    # Restore initial gain setting
    time.sleep(5)
    caput('XF:23ID1-OP{FBck}Sts:FB-Sel',1) #Feedback ready
    caput('XF:23ID1-ES{FCCD}cam1:FCRICGain', initial_gain)#, wait=False)
    time.sleep(3)
    print('\nReturned to initial gain setting and restarting acquisition')

    # Restart camera acquisition
    caput('XF:23ID1-ES{FCCD}cam1:Acquire', 1)#, wait=False)

    # Open shutter again if it was initially open
    if shutter_open:
        shopen()
        time.sleep(0.5)
```

**Comment by danielballan** [comment]
Sorry for the very delayed response. Indeed, the only problem here (the RE being marked in a 'running' state even though it is in fact idle) has been fixed since 0.5.3, the version you are currently running. When you upgrade, you shouldn't see this problem again.
